### PR TITLE
Substitution map API cleanup

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@swiftix Since you're the primary consumer of SubstitutionMaps, you should review this :-)

I instated the stricter assertion and fixed the devirtualizer issue. Unfortunately, the change from GenericSignature::getSubstitutions() to SubstitutionMap::toList() introduced a new counterexample in partial specialization.

I think this means that you had a SubstitutionMap that was being constructed with one signature, and then converted into a SubstitutionList using another. I'll look at it soon.

Note that checking this in is not urgent because the devirtualizer bug only came up with devirtualizing default witness methods, where the conformance is always abstract, so I believe it was harmless (of course the assert caught the other SIL combiner bug so it's a good idea to put it in eventually).

File: include/swift/SIL/TypeSubstCloner.h [hunk 1]
```diff
@@ -149,19 +149,9 @@ class TypeSubstCloner : public SILClonerWithScopes<ImplClass> {
...
-                  ->getSubstitutionMap(Inst->getSubstitutions()),
-              TempSubstList);
-        }
-        for (auto &Sub : TempSubstList) {
-          Sub = asImpl().getOpSubstitution(Sub);
-        }
+        for (auto Sub : Inst->getSubstitutions())
+          TempSubstList.push_back(asImpl().getOpSubstitution(Sub));
+
         SubstitutionList Subs = TempSubstList;
```
Comment (swiftix) [inline include/swift/SIL/TypeSubstCloner.h]: I'm not sure about this change. The target function (i.e. the one where the cloned instructions are added) may have a different generic signature, e.g. when we perform a partial specialization. Your change assumes that the number of substitutions is always the same for the original and the specialized functions, which is wrong, IMHO.

File: include/swift/AST/SubstitutionMap.h [hunk 2]
```diff
@@ -99,6 +97,9 @@ class SubstitutionMap {
   SubstitutionMap subst(TypeSubstitutionFn subs,
                         LookupConformanceFn conformances) const;
 
+  /// Build an array of substitutions from a substitution map.
+  void toList(SmallVectorImpl<Substitution> &subs) const;
+
```
Comment (swiftix) [inline include/swift/AST/SubstitutionMap.h]: I think we also want a variant `toList(GenericSignature *, SmallVectorImpl<Substitution> &subs)`. Sometimes you indeed build the map using the original generic signature, but produce a result for the new specialized generic signature which is "compatible", but may have less substitutable generic parameters.

Also, may be there should be a typedef for `SmallVectorImpl<Substitution>`? May be`SusbtitutionListImpl`?

File: lib/AST/SubstitutionMap.cpp [hunk 3]
```diff
@@ -226,6 +223,46 @@ SubstitutionMap SubstitutionMap::subst(TypeSubstitutionFn subs,
   return result;
 }
 
+void SubstitutionMap::toList(SmallVectorImpl<Substitution> &result) const {
+  auto *genericSig = getGenericSignature();
+  if (genericSig == nullptr)
+    return;
```
Comment (swiftix) [inline lib/AST/SubstitutionMap.cpp]: To be on the safe side, do `result.clear()` before returning. We don't know if `result` was empty before the call.

File: lib/AST/Type.cpp [hunk 4]
```diff
@@ -3015,12 +3015,7 @@ static Type substType(Type derivedType,
...
 
       SmallVector<Substitution, 4> newSubs;
-      boxTy->getLayout()->getGenericSignature()
-        ->getSubstitutions(subMap, newSubs);
-      for (auto &arg : newSubs) {
-        arg = Substitution(arg.getReplacement()->getCanonicalType(),
-                           arg.getConformances());
-      }
+      subMap.toList(newSubs);
       return SILBoxType::get(boxTy->getASTContext(),
```
Comment (swiftix) [inline lib/AST/Type.cpp]: The original code here performs canonicalization of the type. Your code does not do it.

File: include/swift/AST/SubstitutionMap.h (hunk 2 above)
Comment (swiftix) [inline include/swift/AST/SubstitutionMap.h]: I think you guess it, but an example of compatible generic signatures could be e.g. `<T, U>` and `<T, U where U == Int>`. Both have the same set of generic type parameters in the same order, but the second generic signature requires only one substitution.

**Comment by swiftix** [comment]
Regarding the witness_method devirtualizaiton. @DougGregor mentioned yesterday that his SubstitutionMap related crash on an external project was most likely due to this bug. I'd suggest that you provide a separate PR to fix just that bug.

**Comment by swiftix** [comment]
I'd also suggest that you provide a separate PR for the assertion, so that it is easy to temporarily revert only that PR should we hit more occurrences of the same bug.

**Pull request closed without merging**